dex/testing/dcr/harness.sh

```diff
@@ -0,0 +1,242 @@
+SESSION="dcr-harness"
+NODES_ROOT=~/dextest/dcr
+RPC_USER="user"
+RPC_PASS="pass"
+ALPHA_WALLET_SEED="b280922d2cffda44648346412c5ec97f429938105003730414f10b01e1402eac"
+BETA_WALLET_SEED="aabbcaabbcaabbcaabbcaabbcaabbcaabbcaabbcaabbcaabbcaabbcaabbcaabbc"
+WALLET_PASS=123
+ALPHA_MINING_ADDR="SspUvSyDGSzvPz2NfdZ5LW15uq6rmuGZyhL"
+BETA_MINING_ADDR="SsiuwSRYvH7pqWmRxFJWR8Vmqc3AWsjmK2Y"
+ALPHA_WALLET_PORT="19567"
+BETA_WALLET_PORT="19568"
+ALPHA_RPC_PORT="19570"
+BETA_RPC_PORT="19569"
+ALPHA_PORT="19571"
+
+# WAIT can be used in a send-keys call along with a `wait-for done` command to
+# wait for process termination.
+WAIT="; tmux wait-for -S done"
+
+if [ -d "${NODES_ROOT}" ]; then
+  rm -R "${NODES_ROOT}"
+fi
+
+echo "Writing node config files"
+mkdir -p "${NODES_ROOT}/alpha"
+mkdir -p "${NODES_ROOT}/beta"
+mkdir -p "${NODES_ROOT}/harness-ctl"
+
+################################################################################
+# Configuration Files
+################################################################################
+
+# Alpha ctl config
+cat > "${NODES_ROOT}/alpha/alpha-ctl.conf" <<EOF
+rpcuser=${RPC_USER}
+rpcpass=${RPC_PASS}
+rpccert=${NODES_ROOT}/alpha/rpc.cert
+rpcserver=127.0.0.1:${ALPHA_WALLET_PORT}
+EOF
+
+# Beta ctl config
+cat > "${NODES_ROOT}/beta/beta-ctl.conf" <<EOF
+rpcuser=${RPC_USER}
+rpcpass=${RPC_PASS}
+rpccert=${NODES_ROOT}/beta/rpc.cert
+rpcserver=127.0.0.1:${BETA_WALLET_PORT}
+EOF
+
+# Alpha wallet config
+cat > "${NODES_ROOT}/alpha/w-alpha.conf" <<EOF
+username=${RPC_USER}
+password=${RPC_PASS}
+cafile=${NODES_ROOT}/alpha/rpc.cert
+logdir=${NODES_ROOT}/alpha/log
+appdata=${NODES_ROOT}/alpha
+simnet=1
+enablevoting=1
+enableticketbuyer=1
+ticketbuyer.limit=4
+pass=${WALLET_PASS}
+rpcconnect=127.0.0.1:${ALPHA_RPC_PORT}
+rpclisten=127.0.0.1:${ALPHA_WALLET_PORT}
+nogrpc=1
+EOF
+
+# Beta wallet config
+cat > "${NODES_ROOT}/beta/w-beta.conf" <<EOF
+username=${RPC_USER}
+password=${RPC_PASS}
+cafile=${NODES_ROOT}/beta/rpc.cert
+logdir=${NODES_ROOT}/beta/log
+appdata=${NODES_ROOT}/beta
+simnet=1
+pass=${WALLET_PASS}
+rpcconnect=127.0.0.1:${BETA_RPC_PORT}
+rpclisten=127.0.0.1:${BETA_WALLET_PORT}
+nogrpc=1
+EOF
+
+################################################################################
+# Control Scripts
+################################################################################
+
+# Beta ctl
+cat > "${NODES_ROOT}/harness-ctl/alpha" <<EOF
+#!/bin/sh
+dcrctl -C ${NODES_ROOT}/alpha/alpha-ctl.conf --wallet \$*
+EOF
+chmod +x "${NODES_ROOT}/harness-ctl/alpha"
+
+# Alpha ctl
+cat > "${NODES_ROOT}/harness-ctl/beta" <<EOF
+#!/bin/sh
+dcrctl -C ${NODES_ROOT}/beta/beta-ctl.conf --wallet \$*
+EOF
+chmod +x "${NODES_ROOT}/harness-ctl/beta"
+
+# Alpha mine script
+cat > "${NODES_ROOT}/harness-ctl/mine-alpha" <<EOF
+#!/bin/sh
+  case \$1 in
+      ''|*[!0-9]*)  ;;
+      *) NUM=\$1 ;;
+  esac
+  for i in \$(seq \$NUM) ; do
+    dcrctl -C ${NODES_ROOT}/alpha/alpha-ctl.conf generate 1
+    sleep 0.5
+  done
```

Comment:
Note: In a perfect world, we could just `dcrctl ... generate N`, but my results with that set up are not great, and I suspect that @dnldd was seeing something similar when he wrote this script. Seems to result in spurious unresponsiveness. It might be worth investigating further and potentially opening issues on dcrd or dcrwallet.

Reply:
We've been hacking around this in all the previous forms of this script.  I'm not sure if there is a dcrd issue though.
https://github.com/decred/dcrdata/blob/master/dev/parallel-nodes.tmux#L123 (which came from davecgh and matheusd before that, they might have an idea about why the delay is needed.)